### DISCUSSION TITLE
M5Stick C Plus - BLE Firmware Disable Backlight?  and Custom Firmware with LVGL + BLE?

### DISCUSSION THREAD

**Discussion by efaden** [arendst/Tasmota#15086] [Q&A] [answered]
Hey All,

Have a M5Stick C Plus I am looking to play around with.  I was hoping to build a LVGL firmware with BLE support, but it seems that it is too large.  So I was discussing on Discord and the approach has two options.  

Option 1 - Use BLE firmware and using axp driver just shut off the backlight.

Option 2 - Use a minimal firmware and custom partitions to allow a BLE + LVGL firmware to be built and installed.

Any help appreciated.

**Comment by s-hadinger** [comment]
For the AXP192, you could first try the driver used in M5StickC, hoping that the wiring in the same in M5StickC Plus.

Download `preinit.be` from here: https://github.com/tasmota/autoconf/blob/main/raw/esp32/M5StickC/preinit.be
and copy to the file system. Then reboot.

Go to the Berry console, you should be able to use the `axp` object.

For example try: `axp.set_displaydimmer(0)` to turn off backlighting or `axp.set_displaydimmer(100)` to set it to max

**Reply by efaden** [reply]
Awsome!  That worked perfectly.

If I wanted to go the other route and do a custom build with LVGL+BLE and custom partitions how would I go about it?

Guessing flash the minimal then modify the partitions and then flash a custom build bin matching those partitions?

**Comment by s-hadinger** [comment, marked as answer]
For custom partition, you first need to download into the filesystem `partition.be` from here:
https://github.com/arendst/Tasmota/blob/development/tasmota/berry/modules/partition.be

Then in the Berry console, type `import partition` (no output is produced). Then go back into your browser, you should now see a "Partition Manager" button. Use the option in the lowest part of the screen to fine-tune the partition.

Caveat: to do so, the active partition needs to be the first one. If it's not, just re-flash the firmware once more, it will use the other slot.

**Reply by efaden** [reply]
I'm assuming I need to modify the build also to be able to allow it to go bigger?

**Reply by efaden** [reply]
Alright.... 

so I made a new partition csv...

`# Name,   Type, SubType, Offset,  Size, Flags
nvs,      data, nvs,     0x9000,  0x5000,
otadata,  data, ota,     0xe000,  0x2000,
app0,     app,  ota_0,   0x10000, 0x270000,
app1,     app,  ota_1,   0x280000, 0x130000,
spiffs,   data, spiffs,  0x3B0000,0x50000,
`

And a new board file

`{
  "build": {
    "arduino":{
      "ldscript": "esp32_out.ld"
    },
    "core": "esp32",
    "extra_flags": "-DARDUINO_ESP32_DEV -DBOARD_HAS_PSRAM -DARDUINO_USB_CDC_ON_BOOT=0 -DESP32_4M",
    "f_cpu": "80000000L",
    "f_flash": "40000000L",
    "flash_mode": "dout",
    "mcu": "esp32",
    "variant": "esp32",
    "partitions": "esp32_partition_app2560k_spiffs320k.csv"
  },
  "connectivity": [
    "wifi",
    "bluetooth",
    "ethernet",
    "can"
  ],
  "debug": {
    "openocd_target": "esp32.cfg"
  },
  "frameworks": [
    "arduino",
    "espidf"
  ],
  "name": "Espressif Generic ESP32 4M Flash, Tasmota 1856k Code/OTA, 320k FS",
  "upload": {
    "flash_size": "4MB",
    "maximum_ram_size": 327680,
    "maximum_size": 4194304,
    "require_upload_port": true,
    "speed": 460800
  },
  "url": "https://en.wikipedia.org/wiki/ESP32",
  "vendor": "Espressif"
}
`

and a new section in cenv.ini

`
[env:tasmota32-cplus]
extends                 = env:tasmota32-lvgl
board                   = esp32-m5stickcplus
build_flags             = ${env:tasmota32-lvgl.build_flags} -DFIRMWARE_BLUETOOTH
lib_extra_dirs          = lib/libesp32, lib/libesp32_div, lib/lib_basic, lib/lib_i2c, lib/lib_ssl, lib/lib_display, lib/libesp32_lvgl`

**Reply by efaden** [reply]
The build now compiles....  

So I'm guessing from here, I flash minimal and make sure I am in the first app partition, resize it to 2560k?  .... do my numbers all add up?  

Then I flash minimal again to switch into the second app partition being active.  Then flash my bin which should go back into the first and reboot into it.

Correct?

**Reply by efaden** [reply]
Also how do I handle OTA updates?  Do I just do them always manually by doing the minimal first, or is there a way I can flag it to automatically use a minimal intermediate?
